File: _legacy/packages/core/src/sync/ingest.ts
import { createHash } from 'crypto';
import { loadConfigAsync } from '../config/config.js';
import { getDb } from '../db/index.js';
import { GitHubClient } from '../github/client.js';
import {
  fetchFileTree,
  fetchOwnedRepos,
  fetchReadme,
  fetchStarredRepos,
  type GitHubRepo,
} from '../github/repos.js';
import { splitIntoChunks } from './chunker.js';

export interface IngestOptions {
  includeOwned: boolean;
  includeStarred: boolean;
  fullTree: boolean; // if true, always fetch file tree regardless of README length
  verbose: boolean;
}

export type IngestProgressEvent =
  | { type: 'repo'; repo: string; status: 'new' | 'updated' }
  | { type: 'skip'; repo: string; reason: 'unchanged' }
  | { type: 'error'; repo: string; error: string }
  | { type: 'done'; total: number; skipped: number; errors: number };

function hashPushedAt(pushedAt: string): string {
  return createHash('sha256').update(pushedAt).digest('hex');
}

export async function* ingestRepos(
  options: IngestOptions
): AsyncGenerator<IngestProgressEvent> {
  // 1. Load config
  const config = await loadConfigAsync();
  if (!config.githubPat) {
    throw new Error('GitHub PAT not configured');
  }

  // 2. Create client
  const client = new GitHubClient(config.githubPat);

  // 3. Open DB
  const db = getDb(config.dbPath);

  // 4. Collect repos
  const reposToProcess = new Map<
    string,
    { repo: GitHubRepo; isOwned: boolean; isStarred: boolean }
  >();

  try {
    if (options.includeOwned) {
      for await (const repo of fetchOwnedRepos(client)) {
        reposToProcess.set(repo.full_name, {
          repo,
          isOwned: true,
          isStarred: false,
        });
      }
    }

    if (options.includeStarred) {
      for await (const repo of fetchStarredRepos(client)) {
        const existing = reposToProcess.get(repo.full_name);
        if (existing) {
          existing.isStarred = true;
        } else {
          reposToProcess.set(repo.full_name, {
            repo,
            isOwned: false,
            isStarred: true,
          });
        }
      }
    }
  } catch (error: unknown) {
    // If fetching lists fails, we can't do much
    throw new Error(`Failed to fetch repositories: ${error instanceof Error ? error.message : error}`);
  }

  // 4b. Record sync start in sync_state
  const syncType = [options.includeOwned ? 'owned' : '', options.includeStarred ? 'starred' : ''].filter(Boolean).join('+');
  const syncStateResult = db.prepare(`
    INSERT INTO sync_state (sync_type, status, total_repos, started_at)
    VALUES (?, 'in_progress', ?, ?)
  `).run(syncType, reposToProcess.size, new Date().toISOString());
  const syncStateId = syncStateResult.lastInsertRowid;

  let skippedCount = 0;
  let errorCount = 0;
  let processedCount = 0;

  try {
    // Prepare statements
    const checkStmt = db.prepare(
      'SELECT pushed_at_hash FROM repos WHERE github_id = ?'
    );
    
    // Note: We only insert columns that exist in the schema
    const upsertRepoStmt = db.prepare(`
      INSERT OR REPLACE INTO repos (
        github_id, owner, name, full_name, description, url, 
        primary_language, topics, stars, is_fork, is_private, 
        is_starred, is_owned, readme, file_tree, pushed_at, 
        pushed_at_hash, synced_at
      ) VALUES (
        @github_id, @owner, @name, @full_name, @description, @url,
        @primary_language, @topics, @stars, @is_fork, @is_private,
        @is_starred, @is_owned, @readme, @file_tree, @pushed_at,
        @pushed_at_hash, @synced_at
      )
    `);

    const deleteChunksStmt = db.prepare('DELETE FROM chunks WHERE repo_id = ?');
    
    const insertChunkStmt = db.prepare(`
      INSERT INTO chunks (repo_id, chunk_type, chunk_index, content)
      VALUES (@repo_id, @chunk_type, @chunk_index, @content)
    `);

    // 5. Process repos
    for (const [fullName, { repo, isOwned, isStarred }] of reposToProcess) {
      try {
        // 5a. Compute hash
        const pushedAtHash = hashPushedAt(repo.pushed_at);

        // 5b. Look up existing
        const existing = checkStmt.get(repo.id) as { pushed_at_hash: string } | undefined;
        const isExisting = !!existing;

        // 5c. Check if unchanged
        if (existing && existing.pushed_at_hash === pushedAtHash) {
          skippedCount++;
          yield { type: 'skip', repo: fullName, reason: 'unchanged' };
          continue;
        }

        // 5d. Fetch additional data
        const readme = await fetchReadme(client, repo.owner.login, repo.name);
        
        let fileTree: string | null = null;
        if (options.fullTree || (readme && readme.length >= 100)) {
          fileTree = await fetchFileTree(
            client,
            repo.owner.login,
            repo.name,
            repo.default_branch
          );
        }

        // 5e. Write to DB
        const transaction = db.transaction(() => {
          // Upsert repo
          const syncedAt = new Date().toISOString();
          upsertRepoStmt.run({
            github_id: repo.id,
            owner: repo.owner.login,
            name: repo.name,
            full_name: repo.full_name,
            description: repo.description,
            url: repo.html_url,
            primary_language: repo.language,
            topics: JSON.stringify(repo.topics || []),
            stars: repo.stargazers_count,
            is_fork: repo.fork ? 1 : 0,
            is_private: repo.private ? 1 : 0,
            is_starred: isStarred ? 1 : 0,
            is_owned: isOwned ? 1 : 0,
            readme: readme,
            file_tree: fileTree,
            pushed_at: repo.pushed_at,
            pushed_at_hash: pushedAtHash,
            synced_at: syncedAt,
          });

          const { lastInsertRowid } = db.prepare('SELECT last_insert_rowid() as lastInsertRowid').get() as { lastInsertRowid: number | bigint };
          const repoId = lastInsertRowid;

          // Delete existing chunks
          deleteChunksStmt.run(repoId);

          // Insert metadata chunk
          const metadata = {
            full_name: repo.full_name,
            description: repo.description,
            language: repo.language,
            stars: repo.stargazers_count,
            forks: repo.forks_count,
            topics: repo.topics || [],
            archived: repo.archived,
            fork: repo.fork,
            default_branch: repo.default_branch,
            pushed_at: repo.pushed_at,
            html_url: repo.html_url,
          };

          insertChunkStmt.run({
            repo_id: repoId,
            chunk_type: 'metadata',
            chunk_index: 0,
            content: JSON.stringify(metadata),
          });

          // Insert readme chunks
          if (readme && readme.trim().length > 0) {
            const readmeChunks = splitIntoChunks(readme);
            readmeChunks.forEach((chunk, index) => {
              insertChunkStmt.run({
                repo_id: repoId,
                chunk_type: 'readme',
                chunk_index: index,
                content: chunk,
              });
            });
          }

          // Insert file_tree chunks
          if (fileTree && fileTree.trim().length > 0) {
            const treeChunks = splitIntoChunks(fileTree);
            treeChunks.forEach((chunk, index) => {
              insertChunkStmt.run({
                repo_id: repoId,
                chunk_type: 'file_tree',
                chunk_index: index,
                content: chunk,
              });
            });
          }
        });

        transaction();

        processedCount++;
        // Update sync state
        db.prepare('UPDATE sync_state SET processed_repos = ? WHERE id = ?').run(processedCount, syncStateId);

        yield {
          type: 'repo',
          repo: fullName,
          status: isExisting ? 'updated' : 'new',
        };
      } catch (error: unknown) {
        errorCount++;
        yield { type: 'error', repo: fullName, error: error instanceof Error ? error.message : String(error) };
        // Continue to next repo
      }
    }

    // Mark sync as completed
    db.prepare(`
      UPDATE sync_state 
      SET status = 'completed', completed_at = ? 
      WHERE id = ?
    `).run(new Date().toISOString(), syncStateId);

  } catch (error: unknown) {
    // Mark sync as failed if a fatal error occurs
    if (syncStateId) {
      db.prepare(`
        UPDATE sync_state 
        SET status = 'failed', error = ? 
        WHERE id = ?
      `).run(error instanceof Error ? error.message : String(error), syncStateId);
    }
    throw error;
  }

  yield {
    type: 'done',
    total: processedCount,
    skipped: skippedCount,
    errors: errorCount,
  };
}

File: internal/sync/ingest.go
// Package sync handles repository ingestion from GitHub.
package sync

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"github.com/hackastak/repog/internal/github"
)

// IngestOptions configures the ingestion pipeline.
type IngestOptions struct {
	IncludeOwned   bool
	IncludeStarred bool
	FullTree       bool
	DB             *sql.DB
	GitHubPAT      string
}

// IngestEvent represents a progress event from the ingestion pipeline.
type IngestEvent struct {
	Type    string // "repo", "skip", "error", "done"
	Repo    string
	Status  string // "new" or "updated" for type=="repo"
	Reason  string // for type=="skip" or "error"
	Total   int
	Skipped int
	Errors  int
}

// repoEntry holds a repo and its ownership flags.
type repoEntry struct {
	Repo      github.Repo
	IsOwned   bool
	IsStarred bool
}

// metadataChunk is the JSON structure for metadata chunks.
type metadataChunk struct {
	FullName      string   `json:"full_name"`
	Description   string   `json:"description"`
	Language      string   `json:"language"`
	Stars         int      `json:"stars"`
	Forks         int      `json:"forks"`
	Topics        []string `json:"topics"`
	Archived      bool     `json:"archived"`
	Fork          bool     `json:"fork"`
	DefaultBranch string   `json:"default_branch"`
	PushedAt      string   `json:"pushed_at"`
	HTMLURL       string   `json:"html_url"`
}

// hashPushedAt computes SHA-256 hash of the pushed_at timestamp.
func hashPushedAt(s string) string {
	sum := sha256.Sum256([]byte(s))
	return fmt.Sprintf("%x", sum)
}

// splitContent splits content into chunks if it exceeds the max size.
// Using 25000 chars (~7500 tokens) to stay safely under OpenAI's 8192 token limit.
// Based on observed ratio: 28000 chars ≈ 8400 tokens, so 25000 chars ≈ 7500 tokens.
// Returns a slice of content chunks.
func splitContent(content string, maxChars int) []string {
	if maxChars <= 0 {
		maxChars = 25000
	}

	if len(content) <= maxChars {
		return []string{content}
	}

	var chunks []string
	for i := 0; i < len(content); i += maxChars {
		end := i + maxChars
		if end > len(content) {
			end = len(content)
		}
		chunks = append(chunks, content[i:end])
	}

	return chunks
}

// IngestRepos runs the full ingestion pipeline.
// Events are sent to the returned channel as repos are processed.
// The channel is closed when ingestion completes.
// Never panics — all errors are sent as IngestEvent{Type: "error"}.
func IngestRepos(ctx context.Context, opts IngestOptions) <-chan IngestEvent {
	eventCh := make(chan IngestEvent, 100)

	go func() {
		defer close(eventCh)

		client := github.NewClient(opts.GitHubPAT)
		reposToProcess := make(map[string]repoEntry)

		// Collect owned repos
		if opts.IncludeOwned {
			repoCh, errCh := github.FetchOwnedRepos(ctx, client)
			for repo := range repoCh {
				reposToProcess[repo.FullName] = repoEntry{
					Repo:    repo,
					IsOwned: true,
				}
			}
			if err := <-errCh; err != nil {
				eventCh <- IngestEvent{Type: "error", Reason: err.Error()}
				return
			}
		}

		// Collect starred repos
		if opts.IncludeStarred {
			repoCh, errCh := github.FetchStarredRepos(ctx, client)
			for repo := range repoCh {
				if existing, ok := reposToProcess[repo.FullName]; ok {
					existing.IsStarred = true
					reposToProcess[repo.FullName] = existing
				} else {
					reposToProcess[repo.FullName] = repoEntry{
						Repo:      repo,
						IsStarred: true,
					}
				}
			}
			if err := <-errCh; err != nil {
				eventCh <- IngestEvent{Type: "error", Reason: err.Error()}
				return
			}
		}

		var totalProcessed, skippedCount, errorCount int

		for fullName, entry := range reposToProcess {
			repo := entry.Repo

			// Compute hash
			pushedAtHash := hashPushedAt(repo.PushedAt)

			// Check existing
			var existingHash sql.NullString
			var existingID int64
			err := opts.DB.QueryRow(
				"SELECT id, pushed_at_hash FROM repos WHERE github_id = ?",
				repo.ID,
			).Scan(&existingID, &existingHash)

			isExisting := err == nil

			// Check if file_tree chunk exists when full-tree is requested
			hasFileTree := false
			if isExisting && opts.FullTree {
				var count int
				err := opts.DB.QueryRow(
					"SELECT COUNT(*) FROM chunks WHERE repo_id = ? AND (chunk_type = 'file_tree' OR chunk_type LIKE 'file_tree_part_%')",
					existingID,
				).Scan(&count)
				hasFileTree = err == nil && count > 0
			}

			// Skip if unchanged (but not if full-tree requested and file_tree is missing)
			if isExisting && existingHash.Valid && existingHash.String == pushedAtHash {
				if !opts.FullTree || hasFileTree {
					skippedCount++
					eventCh <- IngestEvent{Type: "skip", Repo: fullName, Reason: "unchanged"}
					continue
				}
			}

			// Fetch README
			readme := github.FetchReadme(ctx, client, repo.Owner.Login, repo.Name)

			// Fetch file tree if needed
			var fileTree string
			if opts.FullTree || len(readme) >= 100 {
				fileTree = github.FetchFileTree(ctx, client, repo.Owner.Login, repo.Name, repo.DefaultBranch)
			}

			// Write to DB in transaction
			tx, err := opts.DB.BeginTx(ctx, nil)
			if err != nil {
				errorCount++
				eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
				continue
			}

			// Upsert repo
			description := ""
			if repo.Description != nil {
				description = *repo.Description
			}
			language := ""
			if repo.Language != nil {
				language = *repo.Language
			}
			topics, _ := json.Marshal(repo.Topics)
			syncedAt := time.Now().UTC().Format(time.RFC3339)

			isOwned := 0
			if entry.IsOwned {
				isOwned = 1
			}
			isStarred := 0
			if entry.IsStarred {
				isStarred = 1
			}
			isPrivate := 0
			if repo.Private {
				isPrivate = 1
			}
			isArchived := 0
			if repo.Archived {
				isArchived = 1
			}
			isFork := 0
			if repo.Fork {
				isFork = 1
			}

			result, err := tx.Exec(`
				INSERT INTO repos (
					github_id, owner, name, full_name, description, language,
					stars, forks, is_private, is_starred, is_owned, is_archived, is_fork,
					pushed_at, pushed_at_hash, topics, html_url, default_branch, synced_at
				) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
				ON CONFLICT(github_id) DO UPDATE SET
					owner = excluded.owner,
					name = excluded.name,
					full_name = excluded.full_name,
					description = excluded.description,
					language = excluded.language,
					stars = excluded.stars,
					forks = excluded.forks,
					is_private = excluded.is_private,
					is_starred = excluded.is_starred,
					is_owned = excluded.is_owned,
					is_archived = excluded.is_archived,
					is_fork = excluded.is_fork,
					pushed_at = excluded.pushed_at,
					pushed_at_hash = excluded.pushed_at_hash,
					topics = excluded.topics,
					html_url = excluded.html_url,
					default_branch = excluded.default_branch,
					synced_at = excluded.synced_at
			`,
				repo.ID, repo.Owner.Login, repo.Name, repo.FullName, description, language,
				repo.StargazersCount, repo.ForksCount, isPrivate, isStarred, isOwned, isArchived, isFork,
				repo.PushedAt, pushedAtHash, string(topics), repo.HTMLURL, repo.DefaultBranch, syncedAt,
			)
			if err != nil {
				_ = tx.Rollback()
				errorCount++
				eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
				continue
			}

			// Get repo ID
			var repoID int64
			if isExisting {
				repoID = existingID
			} else {
				repoID, _ = result.LastInsertId()
			}

			// Delete existing chunks
			_, err = tx.Exec("DELETE FROM chunks WHERE repo_id = ?", repoID)
			if err != nil {
				_ = tx.Rollback()
				errorCount++
				eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
				continue
			}

			// Insert metadata chunk
			metadata := metadataChunk{
				FullName:      repo.FullName,
				Description:   description,
				Language:      language,
				Stars:         repo.StargazersCount,
				Forks:         repo.ForksCount,
				Topics:        repo.Topics,
				Archived:      repo.Archived,
				Fork:          repo.Fork,
				DefaultBranch: repo.DefaultBranch,
				PushedAt:      repo.PushedAt,
				HTMLURL:       repo.HTMLURL,
			}
			metadataJSON, _ := json.Marshal(metadata)

			_, err = tx.Exec(
				"INSERT INTO chunks (repo_id, chunk_type, content) VALUES (?, 'metadata', ?)",
				repoID, string(metadataJSON),
			)
			if err != nil {
				_ = tx.Rollback()
				errorCount++
				eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
				continue
			}

			// Insert readme chunk(s) if present
			// Split into multiple chunks if too large
			if readme != "" {
				readmeChunks := splitContent(readme, 25000)
				readmeSuccess := true
				for i, chunk := range readmeChunks {
					chunkType := "readme"
					if len(readmeChunks) > 1 {
						chunkType = fmt.Sprintf("readme_part_%d", i+1)
					}
					_, err = tx.Exec(
						"INSERT INTO chunks (repo_id, chunk_type, content) VALUES (?, ?, ?)",
						repoID, chunkType, chunk,
					)
					if err != nil {
						_ = tx.Rollback()
						errorCount++
						eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
						readmeSuccess = false
						break
					}
				}
				if !readmeSuccess {
					continue
				}
			}

			// Insert file_tree chunk(s) if present
			// Split into multiple chunks if too large
			if fileTree != "" {
				fileTreeChunks := splitContent(fileTree, 25000)
				fileTreeSuccess := true
				for i, chunk := range fileTreeChunks {
					chunkType := "file_tree"
					if len(fileTreeChunks) > 1 {
						chunkType = fmt.Sprintf("file_tree_part_%d", i+1)
					}
					_, err = tx.Exec(
						"INSERT INTO chunks (repo_id, chunk_type, content) VALUES (?, ?, ?)",
						repoID, chunkType, chunk,
					)
					if err != nil {
						_ = tx.Rollback()
						errorCount++
						eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
						fileTreeSuccess = false
						break
					}
				}
				if !fileTreeSuccess {
					continue
				}
			}

			// Insert sync_state
			_, err = tx.Exec(`
				INSERT INTO sync_state (repo_id, status, last_synced_at)
				VALUES (?, 'completed', ?)
				ON CONFLICT(repo_id) DO UPDATE SET
					status = excluded.status,
					last_synced_at = excluded.last_synced_at,
					error_message = NULL
			`, repoID, syncedAt)
			if err != nil {
				_ = tx.Rollback()
				errorCount++
				eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
				continue
			}

			if err := tx.Commit(); err != nil {
				errorCount++
				eventCh <- IngestEvent{Type: "error", Repo: fullName, Reason: err.Error()}
				continue
			}

			totalProcessed++
			status := "new"
			if isExisting {
				status = "updated"
			}
			eventCh <- IngestEvent{Type: "repo", Repo: fullName, Status: status}
		}

		eventCh <- IngestEvent{
			Type:    "done",
			Total:   totalProcessed,
			Skipped: skippedCount,
			Errors:  errorCount,
		}
	}()

	return eventCh
}
